### PR TITLE
Add Azure Quantum submitter factory

### PR DESCRIPTION
**Pull request by samarsha**

### CODE REVIEW

File: src/Simulation/EntryPointDriver/Azure/Azure.cs [hunk 1]
```diff
@@ -271,26 +270,24 @@ private static void Log(AzureSettings settings, string message)
...
         /// <returns>A QIR submitter.</returns>
-        /// <exception cref="ArgumentNullException"><paramref name="settings"/>.Target is null.</exception>
         private static IQirSubmitter? QirSubmitter(AzureSettings settings) => settings.Target switch
         {
-            null => throw new ArgumentNullException(nameof(settings), "Target is null"),
+            null => null,
             NoOpQirSubmitter.Target => new NoOpQirSubmitter(),
             NoOpSubmitter.Target => new NoOpSubmitter(),
-            _ => null // TODO: Add a factory to Microsoft.Azure.Quantum.Client.
+            _ => SubmitterFactory.QirSubmitter("TODO", settings.Target, settings.CreateWorkspace(), settings.Storage)
```
Comment (samarsha) [inline src/Simulation/EntryPointDriver/Azure/Azure.cs:290]: @cesarzc, `AzureQirSubmitter` in qdk-private has a `providerId` parameter in its constructor, but I'm not sure what to put there. There isn't any provider option in the existing command-line options for Azure submission.

Comment (samarsha) [reply src/Simulation/EntryPointDriver/Azure/Azure.cs:290]: This was based on our conversation where you said the providers would have the same constructor shape as `AzureQirSubmitter` - but it seems like the provider classes won't need their own provider ID, so will their constructors be:

```csharp
string target
IWorkspace workspace
string? storageConnectionString
```

instead?

Comment (cesarzc) [reply src/Simulation/EntryPointDriver/Azure/Azure.cs:290]: Sorry I missed that. Yes, there should be another class that uses `AzureQirSubmitter` but does not expose the `providerId` in its constructor. I'll create a `MicrosoftSubmitter` class following the same pattern that we have for IonQ and Honeywell. I'll open a PR shortly.

File: src/Azure/Azure.Quantum.Client/SubmitterFactory.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    /// Creates Azure Quantum submitters for specific targets.
+    /// </summary>
+    public static class SubmitterFactory
+    {
+        /// <summary>
+        /// Information about each supported QIR submitter.
+        /// </summary>
+        private static readonly ImmutableList<SubmitterInfo> QirSubmitters = ImmutableList.Create(
+            new SubmitterInfo(
+                "microsoft",
```
Comment (samarsha) [inline src/Azure/Azure.Quantum.Client/SubmitterFactory.cs:25]: @cesarzc, do you know what the target prefix for MicrosoftSimulatorSubmitter should be? I guessed `microsoft.*`, but maybe it's `microsoft.simulator.*` or something else.

Comment (cesarzc) [reply src/Azure/Azure.Quantum.Client/SubmitterFactory.cs:25]: Currently it is `microsoft.*`, and for the full state simulator the target name is currently `microsoft.simulator`.

Comment (cesarzc) [reply src/Azure/Azure.Quantum.Client/SubmitterFactory.cs]: This is just the current name that the service team is currently working with. We are coordinating with them to come up with a naming convention that makes sense.

File: src/Azure/Azure.Quantum.Client/SubmitterFactory.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    /// </summary>
+    public static class SubmitterFactory
+    {
+        /// <summary>
+        /// Information about each supported QIR submitter.
+        /// </summary>
+        private static readonly ImmutableList<SubmitterInfo> QirSubmitters = ImmutableList.Create(
+            new SubmitterInfo(
+                new Regex(@"^microsoft\.simulator$"),
+                "Microsoft.Quantum.Providers.Targets.MicrosoftSimulatorSubmitter, Microsoft.Quantum.Providers.Core",
```
Comment (cesarzc) [inline src/Azure/Azure.Quantum.Client/SubmitterFactory.cs:27]: Is the first part (before the `,`) the name of the type and the second part the assembly name? Is this how fully-qualified names are defined?

**Review by cesarzc — APPROVED** [review]

File: src/Azure/Azure.Quantum.Client/SubmitterFactory.cs (hunk 3 above)
Comment (samarsha) [reply src/Azure/Azure.Quantum.Client/SubmitterFactory.cs:27]: Yes, it's an [assembly-qualified name](https://docs.microsoft.com/en-us/dotnet/api/system.type.assemblyqualifiedname?view=netcore-3.1).

**Pull request merged**